### PR TITLE
Docs: Visible type information.

### PR DESCRIPTION
**Pull request by donmccurdy**
This came up in an issue at one point, so thought I'd suggest a way of displaying type information in the docs. ~~This just covers BufferGeometry, but I can update other pages if the approach is agreeable.~~ PR now covers all pages. Method/property names would now use their own permalinks, with type information in a low-contrast color afterward.

[Demo — All docs](https://cdn.rawgit.com/donmccurdy/three.js/docs-visible-types/docs/#api/core/BufferGeometry)

| Screenshot|
|--|
| screen shot 2018-02-12 at 7 43 37 pm|

### CODE REVIEW

**Comment by mrdoob** [comment]
That looks great to me! 👍👍👍

**Comment by WestLangley** [comment]
I do not think the foreground/background contrast, as proposed, is sufficiently legible.

There are [metrics](https://www.w3.org/WAI/WCAG20/quickref/#qr-visual-audio-contrast-contrast) for these things, but I expect someone could select reasonably-high-contrast colors and they would be fine.

**Comment by donmccurdy** [comment]
Pushed an update with higher contrast, matching the comments in our code blocks.

**Comment by WestLangley** [comment]
>Pushed an update with higher contrast

I think a different color scheme for the entire document is in order, quite frankly. I vote for clearly legible.

**Comment by donmccurdy** [comment]
Let's address that in another PR — showing type information at all, with the same legibility as example comments (slightly more legible, actually, since the background is lighter), is an improvement on hover-text-only, especially for touchscreen users. 🙂

**Comment by looeee** [comment]
This is awesome! 😄

**Comment by donmccurdy** [comment]
A few prayers to the regex gods later, this PR should cover all of the docs now. 😅

**Comment by takahirox** [comment]
I'm sure this change is very helpful to users.

One thing, "returning `null` from a method" for the methods returning nothing would be a bit misleading.  So should be like this?

    .dispose () : nothing
    .dispose () : undefined
    .dispose () :

This seems existing issue, so another PR would be good.

**Comment by donmccurdy** [comment]
Good point. Hiding the return type when there is no return value seems cleanest. Updated.

File: docs/page.js [hunk 1]
```diff
@@ -61,6 +61,7 @@ function onDocumentLoad( event ) {
 
 	text = text.replace( /\[(member|property|method|param):([\w]+)\]/gi, "[$1:$2 $2]" ); // [member:name] to [member:name title]
 	text = text.replace( /\[(?:member|property|method):([\w]+) ([\w\.\s]+)\]\s*(\(.*\))?/gi, "<a onclick=\"window.parent.setUrlFragment('" + name + ".$2')\" target=\"_parent\" title=\"" + name + ".$2\" class=\"permalink\">#</a> .<a onclick=\"window.parent.setUrlFragment('" + name + ".$2')\" id=\"$2\">$2</a> $3 : <a class=\"param\" onclick=\"window.parent.setUrlFragment('$1')\">$1</a>" );
+	text = text.replace( / : <a class="param" onclick="window.parent.setUrlFragment\('\w+'\)">null<\/a>/gi, '' ); // remove null return types
```
Comment (mrdoob) [inline docs/page.js]: Hmm... Wouldn't it be better to get id of the `: null` from the source instead?

Comment (mrdoob) [reply docs/page.js]: I do not understand... What does this line do?

**Comment by Mugen87** [comment]
One minor thing. It seems that the syntax highlighting is broken:
image

It should look like this:

image

File: docs/page.js [hunk 2]
```diff
@@ -59,8 +59,10 @@ function onDocumentLoad( event ) {
...
 
-	text = text.replace( /\[(?:member|property|method):([\w]+)\]/gi, "[member:$1 $1]" ); // [member:name] to [member:name title]
-	text = text.replace( /\[(?:member|property|method):([\w]+) ([\w\.\s]+)\]/gi, "<a onclick=\"window.parent.setUrlFragment('" + name + ".$2')\" target=\"_parent\" title=\"" + name + ".$2\" class=\"permalink\">#</a> .<a onclick=\"window.parent.setUrlFragment('$1')\" title=\"$1\" id=\"$2\">$2</a> " );
+	text = text.replace( /\[(member|property|method|param):([\w]+)\]/gi, "[$1:$2 $2]" ); // [member:name] to [member:name title]
+	text = text.replace( /\[(?:member|property|method):([\w]+) ([\w\.\s]+)\]\s*(\(.*\))?/gi, "<a onclick=\"window.parent.setUrlFragment('" + name + ".$2')\" target=\"_parent\" title=\"" + name + ".$2\" class=\"permalink\">#</a> .<a onclick=\"window.parent.setUrlFragment('" + name + ".$2')\" id=\"$2\">$2</a> $3 : <a class=\"param\" onclick=\"window.parent.setUrlFragment('$1')\">$1</a>" );
+	text = text.replace( / : <a class="param" onclick="window.parent.setUrlFragment\('\w+'\)">null<\/a>/gi, '' ); // remove null return types
```
Comment (mrdoob) [inline docs/page.js]: I don't think we need this line. What's wrong with showing `: null`?

**Comment by donmccurdy** [comment]
> One minor thing. It seems that syntax highlighting is broken:

Strange... I think that's caused by Rawgit — there are errors on the console about MIME types of the code highlighting CSS. It works when serving the `dev` branch, but perhaps can't do the same from secondary branches? E.g. this [older branch shows the same issue](https://cdn.rawgit.com/donmccurdy/three.js/docs-metal-rough/docs/?q=buffergeometry#Reference/Core/BufferGeometry).

***

@mrdoob I think you left a comment about handling `null` returns more simply, but I can't find it inline, maybe GitHub is struggling with the large diff... Just guessing here, we could change all `[member:name title]` entries inline to something more like:

```
[member:title] () : [return:Foo]
```

... where the `: [return:Foo]` portion is omitted for methods returning null. I think that would result in 1 fewer lines of code, because the `[member:name] to [member:name title]` regex could be removed. Preference?

**Comment by mrdoob** [comment]
When a method return `null` tends to have a reason. If anything... I would hide the return part when a method returns `undefined` as that is the default.

**Comment by donmccurdy** [comment]
Codebase has 357 matches for `[method:null` and none for `[methed:undefined`. I think we've used `null` in all of the cases where nothing is returned, too, so I can't differentiate here without going through method by method. 😕

**Comment by mrdoob** [comment]
> But, as I said, we can leaving `: null` so far and fix that in another PR (if necessary). 

Yes please.

File: docs/page.js (hunk 2 above)
Comment (mrdoob) [reply docs/page.js]: Lets remove this line and the PR will be good to go.

**Comment by donmccurdy** [comment]
Sounds good, removed the line. Glad to consider other changes in another PR. 👍

File: docs/page.js (hunk 2 above)
Comment (donmccurdy) [reply docs/page.js]: ✅

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
After this PR, I guess the `#` link in the right side of every method is no longer needed?

**Comment by donmccurdy** [comment]
Not functionally necessary, no. I like the pattern where the `#` appears to the left when you hover over a title, just as a clue it's a permalink. But either way.

**Pull request merged**